File: web/src/api/roadmap.ts
import type { LessonExercise } from './lessons';
import { request } from './client';

/**
 * Дорожная карта сербского языка.
 *
 * Карта открыта всем: вопрос «что учить дальше» задают до регистрации. Отметки
 * и проценты появляются только у вошедшего, поэтому запросы карты помечены
 * anonymous — гость видит ту же карту, только без своего прогресса.
 */

export const ROADMAP_LEVELS = ['A1', 'A2', 'B1', 'B2', 'C1', 'C2'] as const;
export type RoadmapLevel = (typeof ROADMAP_LEVELS)[number];

export const ROADMAP_CATEGORIES = [
  'reading',
  'grammar',
  'vocabulary',
  'writing',
] as const;
export type RoadmapCategoryKey = (typeof ROADMAP_CATEGORIES)[number];

export interface RoadmapCategory {
  key: RoadmapCategoryKey;
  title: string;
  local: string;
  about: string;
  /** Раздел, которого ещё нет. В зачёт уровня не идёт. */
  planned?: boolean;
}

export interface RoadmapProgress {
  done: number;
  total: number;
  /** Доля 0..1. Считается на сервере: два независимых деления разошлись бы. */
  ratio: number;
  passed: boolean;
}

export interface RoadmapLevelView {
  level: RoadmapLevel;
  name: string;
  categories: Record<string, RoadmapProgress>;
  passed: boolean;
}

export interface RoadmapOverview {
  levels: RoadmapLevelView[];
  categories: RoadmapCategory[];
  /** Цель: к какому уровню человек идёт. Пусто — цель не выбрана. */
  target: string;
  /** Уровень аккаунта: где человек сейчас. Это не то же, что цель. */
  current: string;
  passingScore: number;
  signedIn: boolean;
}

export interface RoadmapItem {
  id: string;
  level: RoadmapLevel;
  category: RoadmapCategoryKey;
  kind: 'book' | 'link' | 'feed_card' | 'text' | 'grammar_topic' | 'lesson';
  title: string;
  summary: string;
  body?: string;
  payload: Record<string, unknown>;
  position: number;
  status: 'draft' | 'published';
  done: boolean;
  updatedAt: string;
}

export interface RoadmapExerciseSet {
  id: string;
  level: RoadmapLevel;
  category: RoadmapCategoryKey;
  itemId?: string;
  title: string;
  content: { exercises?: LessonExercise[] };
  position: number;
  status: 'draft' | 'published';
  done: boolean;
  score: number;
  updatedAt: string;
}

export interface RoadmapWord {
  id: string;
  level: RoadmapLevel;
  theme: string;
  lemma: string;
  translation: string;
  pos?: string;
  note?: string;
  example?: string;
  rank?: number;
  position: number;
  status: 'draft' | 'published';
  known: boolean;
}

export interface RoadmapSection {
  level: RoadmapLevel;
  category: RoadmapCategory;
  intro: string;
  items: RoadmapItem[];
  exercises: RoadmapExerciseSet[];
  words: RoadmapWord[];
  progress: RoadmapProgress;
}

export interface RoadmapComment {
  id: string;
  level: RoadmapLevel;
  parentId?: string;
  userId: string;
  author: string;
  body: string;
  createdAt: string;
  mine: boolean;
}

export function loadRoadmap(signal?: AbortSignal) {
  // Ручка открыта гостям, но вошедшему всё равно нужен Authorization: без
  // него сервер не знает ни выбранную цель, ни отметки о пройденном.
  return request<RoadmapOverview>('/v1/roadmap', { signal });
}

export function loadRoadmapSection(
  level: string,
  category: string,
  signal?: AbortSignal,
) {
  return request<RoadmapSection>(`/v1/roadmap/${level}/${category}`, {
    signal,
  }).then((section) => ({
    ...section,
    // Older deployments serialized untouched Go slices as null. Treat that as
    // an empty section so a cached response cannot take down the whole page.
    items: section.items ?? [],
    exercises: section.exercises ?? [],
    words: section.words ?? [],
  }));
}

/** Отмечает пункт, набор упражнений или слово. done: false снимает отметку. */
export function markRoadmapDone(
  kind: 'item' | 'exercise' | 'word',
  id: string,
  done: boolean,
  score = 1,
  source: 'manual' | 'trainer' = 'manual',
) {
  return request<{ done: boolean; score?: number }>('/v1/roadmap/progress', {
    method: 'POST',
    body: { kind, id, done, score, source },
  });
}

export function saveRoadmapTarget(level: string) {
  return request<{ target: string }>('/v1/roadmap/target', {
    method: 'PUT',
    body: { level },
  });
}

export function loadRoadmapComments(level: string, signal?: AbortSignal) {
  return request<{ comments: RoadmapComment[] }>(
    `/v1/roadmap/${level}/comments`,
    { signal },
  ).then((response) => response.comments);
}

export function addRoadmapComment(
  level: string,
  body: string,
  parentId?: string,
) {
  return request<RoadmapComment>(`/v1/roadmap/${level}/comments`, {
    method: 'POST',
    body: { body, parentId: parentId ?? '' },
  });
}

export function deleteRoadmapComment(id: string) {
  return request<void>(`/v1/roadmap/comments/${id}`, { method: 'DELETE' });
}

/**
 * Уровень взят целиком: по всем считаемым разделам не ниже порога.
 *
 * Планируемые разделы пропускаются — требовать 80% от Writing, которого ещё
 * нет, значило бы закрыть переход на всех уровнях сразу. Правило то же, что на
 * сервере (roadmap.LevelPassed); здесь оно нужно, чтобы не ждать ответа ради
 * уже известного.
 */
export function levelPassed(
  level: RoadmapLevelView,
  categories: RoadmapCategory[],
): boolean {
  const counted = categories.filter((category) => !category.planned);
  if (counted.length === 0) return false;
  return counted.every((category) => {
    const progress = level.categories[category.key];
    return Boolean(progress && progress.total > 0 && progress.passed);
  });
}

/** Следующая ступень. Пусто, если дальше некуда. */
export function nextLevel(level: string): string {
  const index = ROADMAP_LEVELS.indexOf(level as RoadmapLevel);
  if (index < 0 || index + 1 >= ROADMAP_LEVELS.length) return '';
  return ROADMAP_LEVELS[index + 1] ?? '';
}

File: frontend/lib/models/roadmap.dart
/// Дорожная карта сербского языка.
///
/// Каркас — шесть уровней и четыре раздела — приходит с сервера вместе с
/// описаниями: одни и те же формулировки на сайте и в приложении, и правятся
/// они в одном месте.
library;

class RoadmapCategory {
  final String key;
  final String title;
  final String local;
  final String about;

  /// Раздел, которого ещё нет. В зачёт уровня не идёт: требовать 80% от того,
  /// чего не существует, значило бы закрыть переход навсегда.
  final bool planned;

  const RoadmapCategory({
    required this.key,
    required this.title,
    required this.local,
    required this.about,
    required this.planned,
  });

  factory RoadmapCategory.fromJson(Map<String, dynamic> j) => RoadmapCategory(
        key: (j['key'] ?? '').toString(),
        title: (j['title'] ?? '').toString(),
        local: (j['local'] ?? '').toString(),
        about: (j['about'] ?? '').toString(),
        planned: j['planned'] == true,
      );
}

class RoadmapProgress {
  final int done;
  final int total;

  /// Доля 0..1. Считает сервер: два независимых деления рано или поздно
  /// разошлись бы, и приложение показывало бы не то же, что сайт.
  final double ratio;
  final bool passed;

  const RoadmapProgress({
    required this.done,
    required this.total,
    required this.ratio,
    required this.passed,
  });

  static const empty =
      RoadmapProgress(done: 0, total: 0, ratio: 0, passed: false);

  factory RoadmapProgress.fromJson(Map<String, dynamic> j) => RoadmapProgress(
        done: (j['done'] as num?)?.toInt() ?? 0,
        total: (j['total'] as num?)?.toInt() ?? 0,
        ratio: (j['ratio'] as num?)?.toDouble() ?? 0,
        passed: j['passed'] == true,
      );
}

class RoadmapLevelView {
  final String level;
  final String name;
  final Map<String, RoadmapProgress> categories;
  final bool passed;

  const RoadmapLevelView({
    required this.level,
    required this.name,
    required this.categories,
    required this.passed,
  });

  factory RoadmapLevelView.fromJson(Map<String, dynamic> j) {
    final raw = (j['categories'] as Map?) ?? const {};
    return RoadmapLevelView(
      level: (j['level'] ?? '').toString(),
      name: (j['name'] ?? '').toString(),
      categories: {
        for (final entry in raw.entries)
          entry.key.toString(): RoadmapProgress.fromJson(
              (entry.value as Map).cast<String, dynamic>()),
      },
      passed: j['passed'] == true,
    );
  }

  RoadmapProgress progressOf(String category) =>
      categories[category] ?? RoadmapProgress.empty;
}

class RoadmapOverview {
  final List<RoadmapLevelView> levels;
  final List<RoadmapCategory> categories;

  /// Цель: к какому уровню человек идёт. Пусто — цель не выбрана.
  final String target;

  /// Уровень аккаунта: где человек сейчас. Это не то же, что цель.
  final String current;
  final double passingScore;
  final bool signedIn;

  const RoadmapOverview({
    required this.levels,
    required this.categories,
    required this.target,
    required this.current,
    required this.passingScore,
    required this.signedIn,
  });

  factory RoadmapOverview.fromJson(Map<String, dynamic> j) => RoadmapOverview(
        levels: [
          for (final item in (j['levels'] as List?) ?? const [])
            RoadmapLevelView.fromJson((item as Map).cast<String, dynamic>()),
        ],
        categories: [
          for (final item in (j['categories'] as List?) ?? const [])
            RoadmapCategory.fromJson((item as Map).cast<String, dynamic>()),
        ],
        target: (j['target'] ?? '').toString(),
        current: (j['current'] ?? '').toString(),
        passingScore: (j['passingScore'] as num?)?.toDouble() ?? 0.8,
        signedIn: j['signedIn'] == true,
      );
}

class RoadmapItem {
  final String id;
  final String kind;
  final String title;
  final String summary;
  final String body;
  final Map<String, dynamic> payload;
  final bool done;

  const RoadmapItem({
    required this.id,
    required this.kind,
    required this.title,
    required this.summary,
    required this.body,
    required this.payload,
    required this.done,
  });

  factory RoadmapItem.fromJson(Map<String, dynamic> j) => RoadmapItem(
        id: (j['id'] ?? '').toString(),
        kind: (j['kind'] ?? '').toString(),
        title: (j['title'] ?? '').toString(),
        summary: (j['summary'] ?? '').toString(),
        body: (j['body'] ?? '').toString(),
        payload: ((j['payload'] as Map?) ?? const {}).cast<String, dynamic>(),
        done: j['done'] == true,
      );

  String get url => (payload['url'] ?? '').toString();
  String get slug => (payload['slug'] ?? '').toString();
  String get bookId => (payload['bookId'] ?? '').toString();
  String get trainerTopicId => (payload['trainerTopicId'] ?? '').toString();
}

class RoadmapExerciseSet {
  final String id;
  final String title;
  final List<Map<String, dynamic>> exercises;
  final bool done;
  final double score;

  const RoadmapExerciseSet({
    required this.id,
    required this.title,
    required this.exercises,
    required this.done,
    required this.score,
  });

  factory RoadmapExerciseSet.fromJson(Map<String, dynamic> j) {
    final content =
        ((j['content'] as Map?) ?? const {}).cast<String, dynamic>();
    return RoadmapExerciseSet(
      id: (j['id'] ?? '').toString(),
      title: (j['title'] ?? '').toString(),
      exercises: [
        for (final item in (content['exercises'] as List?) ?? const [])
          (item as Map).cast<String, dynamic>(),
      ],
      done: j['done'] == true,
      score: (j['score'] as num?)?.toDouble() ?? 0,
    );
  }
}

class RoadmapWord {
  final String id;
  final String theme;
  final String lemma;
  final String translation;
  final String note;

  /// Фраза с этим словом и её перевод. Слово в обеих помечено звёздочками:
  /// в сербской фразе оно стоит в падеже, а в русском переводе искать его
  /// иначе нечем — русской морфологии у нас нет.
  final String example;
  final String exampleTranslation;
  final bool known;

  const RoadmapWord({
    required this.id,
    required this.theme,
    required this.lemma,
    required this.translation,
    required this.note,
    required this.example,
    required this.exampleTranslation,
    required this.known,
  });

  factory RoadmapWord.fromJson(Map<String, dynamic> j) => RoadmapWord(
        id: (j['id'] ?? '').toString(),
        theme: (j['theme'] ?? '').toString(),
        lemma: (j['lemma'] ?? '').toString(),
        translation: (j['translation'] ?? '').toString(),
        note: (j['note'] ?? '').toString(),
        example: (j['example'] ?? '').toString(),
        exampleTranslation: (j['exampleTranslation'] ?? '').toString(),
        known: j['known'] == true,
      );
}

/// Кусок размеченной фразы: обычный текст либо само слово.
class ExamplePart {
  final String text;
  final bool target;

  const ExamplePart(this.text, {required this.target});
}

final _exampleMark = RegExp(r'\*([^*]+)\*');

/// Разбирает размеченную фразу на части.
///
/// Помечено ровно одно место — так пишет сборщик примеров и так проверяет его
/// валидатор. Без разметки фраза возвращается целиком: примеры, добавленные
/// автором руками, должны показываться, а не пропадать.
List<ExamplePart> splitExample(String phrase) {
  final source = phrase.trim();
  if (source.isEmpty) return const [];
  final match = _exampleMark.firstMatch(source);
  if (match == null) return [ExamplePart(source, target: false)];

  final parts = <ExamplePart>[];
  final before = source.substring(0, match.start);
  final after = source.substring(match.end);
  if (before.isNotEmpty) parts.add(ExamplePart(before, target: false));
  parts.add(ExamplePart(match.group(1) ?? '', target: true));
  if (after.isNotEmpty) parts.add(ExamplePart(after, target: false));
  return parts;
}

class RoadmapSection {
  final String level;
  final RoadmapCategory category;
  final String intro;
  final List<RoadmapItem> items;
  final List<RoadmapExerciseSet> exercises;
  final List<RoadmapWord> words;
  final RoadmapProgress progress;

  const RoadmapSection({
    required this.level,
    required this.category,
    required this.intro,
    required this.items,
    required this.exercises,
    required this.words,
    required this.progress,
  });

  factory RoadmapSection.fromJson(Map<String, dynamic> j) => RoadmapSection(
        level: (j['level'] ?? '').toString(),
        category: RoadmapCategory.fromJson(
            ((j['category'] as Map?) ?? const {}).cast<String, dynamic>()),
        intro: (j['intro'] ?? '').toString(),
        items: [
          for (final item in (j['items'] as List?) ?? const [])
            RoadmapItem.fromJson((item as Map).cast<String, dynamic>()),
        ],
        exercises: [
          for (final item in (j['exercises'] as List?) ?? const [])
            RoadmapExerciseSet.fromJson((item as Map).cast<String, dynamic>()),
        ],
        words: [
          for (final item in (j['words'] as List?) ?? const [])
            RoadmapWord.fromJson((item as Map).cast<String, dynamic>()),
        ],
        progress: RoadmapProgress.fromJson(
            ((j['progress'] as Map?) ?? const {}).cast<String, dynamic>()),
      );

  bool get isEmpty => items.isEmpty && exercises.isEmpty && words.isEmpty;
}

class RoadmapComment {
  final String id;
  final String parentId;
  final String author;
  final String body;
  final DateTime createdAt;
  final bool mine;

  const RoadmapComment({
    required this.id,
    required this.parentId,
    required this.author,
    required this.body,
    required this.createdAt,
    required this.mine,
  });

  factory RoadmapComment.fromJson(Map<String, dynamic> j) => RoadmapComment(
        id: (j['id'] ?? '').toString(),
        parentId: (j['parentId'] ?? '').toString(),
        author: (j['author'] ?? 'Читатель').toString(),
        body: (j['body'] ?? '').toString(),
        createdAt:
            DateTime.tryParse((j['createdAt'] ?? '').toString())?.toLocal() ??
                DateTime.now(),
        mine: j['mine'] == true,
      );
}

/// Уровень взят целиком: по всем считаемым разделам не ниже порога.
///
/// Планируемые разделы пропускаются. Правило то же, что на сервере
/// (roadmap.LevelPassed) — здесь оно нужно, чтобы не ждать ответа ради уже
/// известного.
bool roadmapLevelPassed(
    RoadmapLevelView level, List<RoadmapCategory> categories) {
  final counted = categories.where((category) => !category.planned).toList();
  if (counted.isEmpty) return false;
  return counted.every((category) {
    final progress = level.progressOf(category.key);
    return progress.total > 0 && progress.passed;
  });
}

/// Следующая ступень. Пусто, если дальше некуда.
String roadmapNextLevel(String level, List<RoadmapLevelView> levels) {
  final index = levels.indexWhere((item) => item.level == level);
  if (index < 0 || index + 1 >= levels.length) return '';
  return levels[index + 1].level;
}
